pkg/ocibundle/tools/overlay_linux.go

```diff
@@ -79,41 +83,175 @@ func CreateOverlayTmpfs(bundlePath string, sizeMiB int) error {
+	}
+	if err := ensureOverlayDir(workSubdirOf(dir), true, 0o700); err != nil {
+		return fmt.Errorf("err encountered while preparing work subdir of overlay dir %q: %w", workSubdirOf(dir), err)
+	}
+
+	return nil
+}
+
+// performIdentityMounts performs the preparator identity mounts of overlay dirs, typically carried out prior to the actual overlay mount where these dirs are used as lowerdir or upperdir arguments.
+func performIdentityMounts(ovs OverlaySet) error {
+	var err error
+
+	locsToBind := ovs.ReadonlyLocs
+	if len(ovs.WritableLoc) > 0 {
+		// Check if writable overlay dir already exists; if it doesn't, try to create it.
+		if err = ensureOverlayDir(ovs.WritableLoc, true, 0o755); err != nil {
+			return err
+		}
+
+		locsToBind = append(locsToBind, ovs.WritableLoc)
+	}
+
+	// Try to do initial bind-mounts
+	for _, d := range locsToBind {
+		if err = ensureOverlayDir(d, false, 0); err != nil {
+			return fmt.Errorf("error accessing directory %s: %s", d, err)
+		}
+
+		if err = syscall.Mount(d, d, "", syscall.MS_BIND, ""); err != nil {
+			return fmt.Errorf("failed to bind %s: %s", d, err)
+		}
+
+		// best effort to cleanup mount
+		defer func() {
+			if err != nil {
+				syscall.Unmount(d, syscall.MNT_DETACH)
+			}
+		}()
+
+		// Try to perform remount
+		if err = syscall.Mount("", d, "", syscall.MS_REMOUNT|syscall.MS_BIND, ""); err != nil {
+			return fmt.Errorf("failed to remount %s: %s", d, err)
+		}
+	}
+
+	return err
+}
+
+// overlayOptions creates the options string to be used in an overlay mount
+func overlayOptions(rootFsDir string, ovs OverlaySet) string {
+	// Create lowerdir argument of options string
+	lowerDirJoined := strings.Join(append(ovs.ReadonlyLocs, rootFsDir), ":")
+
+	if len(ovs.WritableLoc) > 0 {
+		return fmt.Sprintf("lowerdir=%s,upperdir=%s,workdir=%s", lowerDirJoined, upperSubdirOf(ovs.WritableLoc), workSubdirOf(ovs.WritableLoc))
+	}
+
+	return fmt.Sprintf("lowerdir=%s", lowerDirJoined)
+}
+
+// performOverlayMount mounts an overlay atop a given rootfs directory
+func performOverlayMount(rootFsDir, options string) error {
+	// Try to perform actual mount
 	if err := syscall.Mount("overlay", rootFsDir, "overlay", 0, options); err != nil {
-		return fmt.Errorf("failed to mount %s: %s", overlayDir, err)
+		return fmt.Errorf("failed to mount %s: %s", rootFsDir, err)
+	}
+
+	return nil
+}
+
+// ensureOverlayDir checks if a directory already exists; if it doesn't, and writable is true, it attempts to create it with the specified permissions.
+func ensureOverlayDir(dir string, createIfMissing bool, createPerm os.FileMode) error {
+	_, err := os.Stat(dir)
+	if len(dir) == 0 {
+		return fmt.Errorf("internal error: ensureOverlayDir() called with empty dir name")
+	}
+
+	if err == nil {
+		return nil
+	}
+
+	if !os.IsNotExist(err) {
+		return err
+	}
+
+	if !createIfMissing {
+		return fmt.Errorf("missing overlay dir %#v", dir)
```

Comment:
Nit (which I think I've overlooked before) - it's a little unusual to use %#v for something that's definitely a string. %q is string specific and equivalent here, afaik.

'%q' is used quite widely in the code while '%#v' is not, so prefer following the existing pattern unless there's a specific reason, i.e. quoted / go syntax output of a non-string type is required.